cdc_iterator.go

```diff
@@ -0,0 +1,243 @@
+		}
+
+		c.acks.Add(1)
+		return rec, nil
+	}
+}
+
+func (c *cdcIterator) Teardown(ctx context.Context) error {
+	close(c.canalDoneC)
+
+	c.canal.Close()
+	if err := <-c.canalRunErrC; err != nil {
+		return fmt.Errorf("failed to stop canal: %w", err)
+	}
+
+	if err := c.acks.Wait(ctx); err != nil {
+		return fmt.Errorf("failed to wait for cdc acks: %w", err)
+	}
+
+	return nil
+}
+
+func buildPayload(columns []schema.TableColumn, rows []any) sdk.StructuredData {
+	payload := sdk.StructuredData{}
+	for i, col := range columns {
+		payload[col.Name] = common.FormatValue(rows[i])
+	}
+	return payload
+}
+
+func (c *cdcIterator) buildRecord(e *canal.RowsEvent) (sdk.Record, error) {
+	pos, err := c.canal.GetMasterPos()
+	if err != nil {
+		return sdk.Record{}, fmt.Errorf("failed to get master position from buildRecord: %w", err)
+	}
+
+	metadata := sdk.Metadata{"mysql.action": e.Action}
+	metadata.SetCollection(e.Table.Name)
+
+	pos.Pos = e.Header.LogPos
+
+	switch e.Action {
+	case canal.InsertAction:
+		position := common.CdcPosition{Position: pos}.ToSDKPosition()
+		payload := buildPayload(e.Table.Columns, e.Rows[0])
+
+		table := common.TableName(e.Table.Name)
+		primaryKey := c.config.TableKeys[table]
+
+		key, err := buildRecordKey(primaryKey, table, e.Action, payload)
+		if err != nil {
+			return sdk.Record{}, fmt.Errorf("failed to build record key: %w", err)
+		}
+
+		return sdk.Util.Source.NewRecordCreate(position, metadata, key, payload), nil
+	case canal.DeleteAction:
+		position := common.CdcPosition{Position: pos}.ToSDKPosition()
+
+		payload := buildPayload(e.Table.Columns, e.Rows[0])
+
+		table := common.TableName(e.Table.Name)
+		primaryKey := c.config.TableKeys[table]
+
+		key, err := buildRecordKey(primaryKey, table, e.Action, payload)
+		if err != nil {
+			return sdk.Record{}, fmt.Errorf("failed to build record key: %w", err)
+		}
+
+		return sdk.Util.Source.NewRecordDelete(position, metadata, key), nil
+	case canal.UpdateAction:
+		position := common.CdcPosition{Position: pos}.ToSDKPosition()
+		before := buildPayload(e.Table.Columns, e.Rows[0])
+		after := buildPayload(e.Table.Columns, e.Rows[1])
+
+		table := common.TableName(e.Table.Name)
+		primaryKey := c.config.TableKeys[table]
+
+		key, err := buildRecordKey(primaryKey, table, e.Action, before)
+		if err != nil {
+			return sdk.Record{}, fmt.Errorf("failed to build record key: %w", err)
+		}
+
+		return sdk.Util.Source.NewRecordUpdate(position, metadata, key, before, after), nil
+	}
+
+	return sdk.Record{}, fmt.Errorf("unknown row event action: %s", e.Action)
+}
+
+func buildRecordKey(
+	primaryKey common.PrimaryKeyName, table common.TableName,
+	action string, payload sdk.StructuredData,
+) (sdk.StructuredData, error) {
+	val, ok := payload[string(primaryKey)]
+	if !ok {
+		return nil, fmt.Errorf("key %s not found in payload", primaryKey)
+	}
+
+	return sdk.StructuredData{
+		string(primaryKey): val,
+		"table":            string(table),
+		"action":           action,
```

Comment:
Ohh, no, please don't do that. The key has to be specifically and _only_ the identifier of the row. Destination connectors use this key to update the row in the destination resource. This makes it seem like the source row contained columns `table` and `action`, which is not the ecase.

The table and action should be added into the record metadata. The table should go into `opencdc.collection` and the `action` is used to pick the right operation. All of this is already done as far as I can see.